### PR TITLE
KFSPTS-15573 adding more award extended attributes

### PR DESCRIPTION
**Pull request by jhulslander**
Please don't merge this yet.  There are some open questions to Janet and I need to do some more local validation.  When taking a look at AwardMaintenanceDocument, please note I couldn't get  AwardMaintenanceDocument-Accounts section to work using the standard data dictionary over rides.   It might be easier to discuss verbally what I was finding.

I am opening this up for code review to get feedback on AwardMaintenanceDocument-Accounts section and the code in general.  Please do NOT merge this code yet.

### CODE REVIEW

File: src/main/resources/edu/cornell/kfs/module/ar/cu-spring-ar.xml [hunk 1]
```diff
@@ -72,6 +73,11 @@
           p:code="13"
           p:financialSystemDocumentService-ref="financialSystemDocumentService"
           p:contractsGrantsInvoiceDocumentDao-ref="contractsGrantsInvoiceDocumentDao"/>
+          
+	<bean id="autoApproveDisabledSuspensionCategory" parent="autoApproveDisabledSuspensionCategory-parentBean"/>
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/ar/cu-spring-ar.xml]: formatting looks a bit off on this line.

File: src/main/resources/edu/cornell/kfs/module/cg/cu-ojb-cg.xml [hunk 2]
```diff
@@ -200,5 +203,53 @@
...
+        <reference-descriptor name="chartOfAccounts" class-ref="org.kuali.kfs.coa.businessobject.Chart"
+                              auto-retrieve="true"
+                              auto-update="none" auto-delete="none" proxy="true">
+            <foreignkey field-ref="chartOfAccountsCode"/>
+        </reference-descriptor>
+        <reference-descriptor name="extension" class-ref="edu.cornell.kfs.module.cg.businessobject.AwardAccountExtendedAttribute" auto-retrieve="true" auto-update="object" auto-delete="object" proxy="false">
+            <foreignkey field-ref="proposalNumber" />
+            <foreignkey field-ref="chartOfAccountsCode" />
+            <foreignkey field-ref="accountNumber" />
+		</reference-descriptor>
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/cu-ojb-cg.xml]: formatting looks off

File: src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardExtendedAttribute.java [hunk 3]
```diff
@@ -98,4 +101,28 @@ public void setEverify(boolean everify) {
...
+
+    public String getAutoApproveReason() {
+        return autoApproveReason;
+    }
+
+    public void setAutoApproveReason(String autoApproveReason) {
+        this.autoApproveReason = autoApproveReason;
+    }
+
+    public String getPurchasOrderNumber() {
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardExtendedAttribute.java]: please check there is a misspelling in purchase

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccount.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:p="http://www.springframework.org/schema/p"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+                http://www.springframework.org/schema/beans/spring-beans-2.0.xsd">
+                
+	<bean id="AwardAccount" parent="AwardAccount-parentBean">
+		<property name="attributes">
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccount.xml]: formatting is a bit off on this file

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+			<list>
+				<ref bean="AwardAccountExtendedAttribute--proposalNumber" />
+				<ref bean="AwardAccountExtendedAttribute-chartOfAccountsCode" />
+				<ref bean="AwardAccountExtendedAttribute-accountNumber" />
+				<ref bean="AwardAccountExtendedAttribute-accountPurchaseOrderNumber" />
+			</list>
+		</property>
+	</bean>
+	
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute--proposalNumber" parent="Award-proposalNumber"/>
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: there is an extra dash line between AwardAccountExtendedAttribute and proposalNumber. Has this maybe cause some issues?

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+				<ref bean="AwardAccountExtendedAttribute-accountPurchaseOrderNumber" />
+			</list>
+		</property>
+	</bean>
+	
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute--proposalNumber" parent="Award-proposalNumber"/> 
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute-chartOfAccountsCode" parent="AwardAccount-chartOfAccountsCode"/> 
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute-accountNumber" parent="AwardAccount-accountNumber"/> 
+	
+	<bean id="AwardAccountExtendedAttribute-accountPurchaseOrderNumber" parent="AwardAccountExtendedAttribute-accountPurchaseOrderNumber-parentBean"/>
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: formatting is a bit off

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml [hunk 7]
```diff
@@ -99,5 +102,22 @@
         </property>
         <property name="formatterClass" value="org.kuali.rice.core.web.format.BooleanFormatter"/>
   	</bean>
+  	
+  	<bean id="AwardExtendedAttribute-primeAgreementNumber" parent="AwardExtendedAttribute-primeAgreementNumber-parentBean"/>
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml]: formatting is a bit off

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	xmlns:dd="http://rice.kuali.org/dd"
+	xsi:schemaLocation=" http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-2.0.xsd http://rice.kuali.org/dd http://rice.kuali.org/dd/dd.xsd">
+
+	<bean id="AwardAccountExtendedAttribute" parent="AwardAccountExtendedAttribute-parentBean" />
+	<bean id="AwardAccountExtendedAttribute-parentBean" abstract="true" parent="BusinessObjectEntry">
+		<property name="businessObjectClass" value="edu.cornell.kfs.module.cg.businessobject.AwardAccountExtendedAttribute" />
+		<property name="objectLabel" value="AwardAccountExtendedAttribute" />
+		<property name="attributes">
+			<list>
+				<ref bean="AwardAccountExtendedAttribute--proposalNumber" />
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: there is an extra dash between AwardAccountExtendedAttribute and proposalNumber. Has this maybe cause some issues?

File: src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardExtendedAttribute.java (hunk 3 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardExtendedAttribute.java]: I will fix

File: src/main/resources/edu/cornell/kfs/module/ar/cu-spring-ar.xml (hunk 1 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/ar/cu-spring-ar.xml]: OK, will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccount.xml (hunk 4 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccount.xml]: OK, will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml (hunk 8 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: ok, will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml (hunk 5 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml (hunk 6 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml (hunk 7 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml]: will fix

File: src/main/resources/edu/cornell/kfs/module/cg/cu-ojb-cg.xml (hunk 2 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/cu-ojb-cg.xml]: OK, will fix

**Comment by jhulslander** [comment]
This is ready for review again.  I believe the formatting has been fixed, and the noted typos have have been corrected.  Janet got back to me on my open questions and I have made some changes to address them, in particular, the new business rule isn't enforced when the KFS system user is initiating the award, this allows the RASS and ezra batch job to create awards.

**Comment by jhulslander** [comment]
Also , I tried the data dictionary over ride again, and the bug I noted in my email is there.

Once you believe this code is ready, I think it can be merged.

File: src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java [hunk 9]
```diff
@@ -229,6 +236,38 @@ protected boolean checkAccountNotUsedOnOtherAwards(AwardAccount awardAccount) {
...
+	    String documentInitiator = document.getDocumentHeader().getWorkflowDocument().getInitiatorPrincipalId();
+	    Collection<String> byPassInitiators = getParameterService().getParameterValuesAsString(KFSConstants.OptionalModuleNamespaces.CONTRACTS_AND_GRANTS, 
+	            CUKFSConstants.CGParms.AWARD_RULE_COMPNONENT, CUKFSConstants.CGParms.BYPASS_AWARD_EXTENSION_AUTO_APPROVE_REASON_RULE_INITIATORS);
+	    if (LOG.isDebugEnabled()) {
+	        LOG.debug("shouldEnforceAutoApproveReasonRule, the document intiator: " + documentInitiator);
+	        LOG.debug("shouldEnforceAutoApproveReasonRule, the bypass intiators: " + byPassInitiators);
+	    }
+	    
+	    if (CollectionUtils.isNotEmpty(byPassInitiators) && byPassInitiators.contains(documentInitiator)) {
+	        LOG.debug("shouldEnforceAutoApproveReasonRule, the document initiator is in the list of bypass initiators, so do not enfource for the auto approve reason rule.");
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java]: there is a missplelling here: enfource

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml [hunk 10]
```diff
@@ -100,4 +103,21 @@
...
+		p:name="autoApproveReason" p:label="Auto Approve Reason" p:shortLabel="Auto Approve Reason" p:maxLength="120"
+		p:validationPattern-ref="AnyCharacterWithWhitespaceValidation">
+		<property name="control">
+			<bean parent="TextareaControlDefinition" p:rows="2" p:cols="60" />
+		</property>
+	</bean>
+
+	<bean id="AwardExtendedAttribute-purchaseOrderNumber" parent="AwardExtendedAttribute-purchaseOrderNumber-parentBean" />
+	<bean id="AwardExtendedAttribute-purchaseOrderNumber-parentBean" abstract="true" parent="Proposal-grantNumber"
+		p:name="purchaseOrderNumber" p:label="Purchas Order #" p:shortLabel="PO #" />
```
Comment (db483) [inline src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml]: another misspelling

File: src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardAccountExtendedAttribute.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    private static final long serialVersionUID = 4036900233631513201L;
+    
+    private String proposalNumber;
+    private String chartOfAccountsCode;
+    private String accountNumber;
+    private String accountPurchaseOrderNumber;
+    
+    public String getProposalNumber() {
+        return proposalNumber;
+    }
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/cg/businessobject/AwardAccountExtendedAttribute.java]: OK, will fix ..... interesting, the eclipse tool to generate getters and setters must not add code per the defined code style.

File: src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java [hunk 12]
```diff
@@ -229,6 +236,38 @@ protected boolean checkAccountNotUsedOnOtherAwards(AwardAccount awardAccount) {
...
+            if (StringUtils.isBlank(awardExtendedAttribute.getAutoApproveReason())) {
+                putFieldError("extension.autoApproveReason", CUKFSKeyConstants.ERROR_AUTO_APPROVE_REASON_REQUIRED);
+                success = false;
+            }
+        }
+        
+        return success;
+    }
+	
+	private boolean shouldEnforceAutoApproveReasonRule(MaintenanceDocument document) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java]: will fix

File: src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java (hunk 9 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/cg/document/validation/impl/AwardExtensionRule.java]: will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+		</property>
+	</bean>
+
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute-proposalNumber" parent="Award-proposalNumber" />
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute-chartOfAccountsCode" parent="AwardAccount-chartOfAccountsCode" />
+	<dd:boAttributeRef id="AwardAccountExtendedAttribute-accountNumber" parent="AwardAccount-accountNumber" />
+
+	<bean id="AwardAccountExtendedAttribute-accountPurchaseOrderNumber" parent="AwardAccountExtendedAttribute-accountPurchaseOrderNumber-parentBean" />
+	<bean id="AwardAccountExtendedAttribute-accountPurchaseOrderNumber-parentBean" abstract="true" parent="Proposal-grantNumber"
+		p:name="accountPurchaseOrderNumber" p:label="Purchas Order #" p:shortLabel="PO #" />
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardAccountExtendedAttribute.xml]: OK, will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml (hunk 10 above)
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/AwardExtendedAttribute.xml]: OK, will fix

File: src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/Award.xml [hunk 14]
```diff
@@ -126,6 +130,26 @@
...
+    <bean id="Award-extension-autoApproveReason-parentBean" abstract="true" parent="AwardExtendedAttribute-autoApproveReason">
+        <property name="name" value="extension.autoApproveReason"/>
+    </bean>
+    
+    <bean id="Award-extension-purchaseOrderNumber" parent="Award-extension-purchaseOrderNumber-parentBean"/>
+    <bean id="Award-extension-purchaseOrderNumber-parentBean" abstract="true" parent="AwardExtendedAttribute-purchaseOrderNumber">
+        <property name="name" value="extension.purchaseOrderNumber"/>
+    </bean>
+    
+    <bean id="Award-AwardAccount-extension-accountPurchaseOrderNumber" parent="Award-AwardAccount-extension-accountPurchaseOrderNumber-parentBean"/>
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/resources/edu/cornell/kfs/module/cg/businessobject/datadictionary/Award.xml]: good catch, turns out it is not needed

**Comment by jhulslander** [comment]
I've made further changes, this is ready for review again

**Pull request merged**